Add log correlation documentation (#481)
signalfx/splunk-otel-go: docs/correlating-traces-and-logs.md

# Correlating Trace and Logs

> The [OpenTelemetry trace API] can be used to extract trace metadata from a
> context containing a span. That metadata can then be used to annotate log
> events so they will be correlated with the trace.

Reviewer: ```suggestion
The [OpenTelemetry trace API] can be used to extract trace metadata from a
context containing a span. That metadata can then be used to annotate log
events so they are correlated with the trace.
```

## Extracting Trace Metadata

> To extract trace metadata, use the [`SpanContextFromContext`] function. This
> will extract the trace metadata from an [`context.Context`] and return it in
> the form of a [`SpanContext`].

Reviewer: ```suggestion
To extract trace metadata, use the [`SpanContextFromContext`] function. This
extracts the trace metadata from an [`context.Context`] and returns it in
the form of a [`SpanContext`].
```

```go
spanContext := trace.SpanContextFromContext(ctx)
if !spanContext.IsValid() {
	// ctx does not contain a valid span.
    // There is no trace metadata to add.
    return
}
```

## Annotating Log Events

> Once you have a [`SpanContext`], you can use the trace metadata it contains to
> annotate log events. How this is done will depend on the logging library used.

Reviewer: ```suggestion
Once you have a [`SpanContext`], you can use the trace metadata to
annotate log events. How this is done depends on the logging library used.
```

> ### Structured Logging

Reviewer: @MrAlias "Structured" as in relying on data structure, or a structured text output, such as TOML/YAML/JSON? I believe Log Observer has an easier time digesting JSON logs.

Author: Here "structured" refers to the data structure and the annotated attributes added to a log event. By contrast, an unstructured logger will usually only accept a log message as a string.

> If you use a structured logger, the trace metadata should be added as logger
> fields.

Reviewer: ```suggestion
If you use a structured logger, add the trace metadata as logger fields.
```

For example, using the [zap] logging library:

```go
logger, _ := zap.NewProduction()
defer logger.Sync()
logger = logger.With(
    zap.String("trace_id", spanContext.TraceID().String()),
    zap.String("span_id", spanContext.SpanID().String()),
    zap.String("trace_flags", spanContext.TraceFlags().String()),
)
logger.Info("failed to fetch URL", zap.String("URL", url))
```

### Unstructured Logging

If using unstructured logging, the trace metadata can be added as a part of the
logged message.

For example using the standard library [`log`] package:

```go
log.Printf(
    "(trace_id: %s, span_id: %s, trace_flags: %s): failed to fetch URL: %s",
    spanContext.TraceID().String(),
    spanContext.SpanID().String(),

>     spanContext.TraceFlags().String(),

Reviewer: It'd be nice to state what metadata is available at the beginning of this document (`TraceID`, `SpanID`, etc.)

    url,
)
```

> Be sure to add the metadata in a predictable manner so it can be parsed later on!

Reviewer: ```suggestion
> Make sure to add the metadata following an order you can parse later on.
```

> ## Example HTTP Server
> 
> Below is a complete example of an HTTP server using the [chi] framework that
> extracts trace metadata and annotates log messages for handled requests.

Reviewer: ```suggestion
## HTTP server example

The following is a complete example of an HTTP server using the [chi] framework that
extracts trace metadata and annotates log messages for handled requests.
```

# Correlating Trace and Logs

The [OpenTelemetry trace API] can be used to extract trace metadata from a
context containing a span. That metadata can then be used to annotate log
events so they are correlated with the trace.

## Extracting Trace Metadata

To extract trace metadata, use the [`SpanContextFromContext`] function. This
extracts the trace metadata from an [`context.Context`] and returns it in
the form of a [`SpanContext`].

```go
spanContext := trace.SpanContextFromContext(ctx)
if !spanContext.IsValid() {
	// ctx does not contain a valid span.

>     // There is no trace metadata to add.

Reviewer 2: something is wrong with indentation here

    return
}
```

The OpenTelemetry [`SpanContext`] contains the trace and span ID, trace flags
that contain sampling information, and tracestate information which holds
vendor specific tracing-system context. All of this information can be added to
log events to enrich their context, but the trace and span ID are the metadata
that must be added to correlate with the trace.

## Annotating Log Events

Once you have a [`SpanContext`], you can use the trace metadata to
annotate log events. How this is done depends on the logging library used.

### Structured Logging

If you use a structured logger, add the trace metadata as logger fields.

For example, using the [zap] logging library:

```go
logger, _ := zap.NewProduction()
defer logger.Sync()
logger = logger.With(
    zap.String("trace_id", spanContext.TraceID().String()),
    zap.String("span_id", spanContext.SpanID().String()),

>     zap.String("trace_flags", spanContext.TraceFlags().String()),

Reviewer 2: spaces or tabs? I think we have tabs in other places

)
logger.Info("failed to fetch URL", zap.String("URL", url))
```

### Unstructured Logging

If using unstructured logging, the trace metadata can be added as a part of the
logged message.

For example using the standard library [`log`] package:

```go
log.Printf(
    "(trace_id: %s, span_id: %s, trace_flags: %s): failed to fetch URL: %s",
    spanContext.TraceID().String(),
    spanContext.SpanID().String(),
    spanContext.TraceFlags().String(),

>     url,

Reviewer 2: spaces or tabs? I think we have tabs in other places
